smart-contracts/contracts/mixins/MixinPurchase.sol

```diff
@@ -57,94 +57,106 @@ contract MixinPurchase is
+      uint idTo = toKey.tokenId;
+      uint newTimeStamp;
+
+      if (idTo == 0) {
+        // Assign a new tokenId (if a new owner or previously transferred)
+        _assignNewTokenId(toKey);
+        // refresh the cached value
+        idTo = toKey.tokenId;
+        _recordOwner(_recipient, idTo);
+        // check for a non-expiring key
+        if (expirationDuration == type(uint).max) {
+          newTimeStamp = type(uint).max;
+        } else {
+          newTimeStamp = block.timestamp + expirationDuration;
+        }
+        toKey.expirationTimestamp = newTimeStamp;
+
+        // set key manager
+        _setKeyManagerOf(idTo, _keyManagers[i]);
+
+        // trigger event
+        emit Transfer(
+          address(0), // This is a creation.
+          _recipient,
+          idTo
+        );
+      } else if (toKey.expirationTimestamp > block.timestamp) {
+        // prevent re-purchase of a valid non-expiring key
+        require(toKey.expirationTimestamp != type(uint).max, 'A valid non-expiring key can not be purchased twice');
+
+        // This is an existing owner trying to extend their key
+        newTimeStamp = toKey.expirationTimestamp + expirationDuration;
+        toKey.expirationTimestamp = newTimeStamp;
+
+        emit RenewKeyPurchase(_recipient, newTimeStamp);
       } else {
-        newTimeStamp = block.timestamp + expirationDuration;
-      }
-      toKey.expirationTimestamp = newTimeStamp;
+        // This is an existing owner trying to renew their expired or cancelled key
+        if(expirationDuration == type(uint).max) {
+          newTimeStamp = type(uint).max;
+        } else {
+          newTimeStamp = block.timestamp + expirationDuration;
+        }
+        toKey.expirationTimestamp = newTimeStamp;
 
-      _setKeyManagerOf(idTo, _keyManager);
+        _setKeyManagerOf(idTo, _keyManagers[i]);
 
-      emit RenewKeyPurchase(_recipient, newTimeStamp);
-    }
+        emit RenewKeyPurchase(_recipient, newTimeStamp);
+      }
 
-    
-    uint inMemoryKeyPrice = _purchasePriceFor(_recipient, _referrer, _data);
+      // price      
+      uint inMemoryKeyPrice = _purchasePriceFor(_recipient, _referrers[i], _data);
+      totalPriceToPay = totalPriceToPay + inMemoryKeyPrice;
+
+      // call Unlock contract to record GNP
+      try unlockProtocol.recordKeyPurchase(inMemoryKeyPrice, _referrers[i]) 
+      {} 
+      catch {
+        // emit missing unlock
+        emit UnlockCallFailed(address(this), address(unlockProtocol));
+      }
 
-    try unlockProtocol.recordKeyPurchase(inMemoryKeyPrice, _referrer) 
-    {} 
-    catch {
-      // emit missing unlock
-      emit UnlockCallFailed(address(this), address(unlockProtocol));
+      if(address(onKeyPurchaseHook) != address(0)) {
+        onKeyPurchaseHook.onKeyPurchase(
+          msg.sender, 
+          _recipient, 
+          _referrers[i], 
+          _data, 
+          inMemoryKeyPrice, 
+          0 // tokenAddress == address(0) ? _value : msg.value
```

Comment:
This is one I am unsure of so far. How is this intended to be used?

Reply:
This is to show the actual price "submitted" by the user (either the `msg.value` or the `_value` for this specific purchase. (user can submit more than required...).

This is why it is important to also make sure the `_value` is in an array... The FE will build the `msg.value` to be the sum of all `_values`.